lib/Transforms/IPO/StructFieldCacheAnalysisImpl.h

```diff
@@ -0,0 +1,346 @@
+  /// It can be allocas/function args/globals
+  void analyzeUsersOfStructPointerValue(const Value *V);
+
+  /// Obtain which field the instruction is accessing and return no val if not
+  /// accessing any struct field
+  Optional<FieldNumType> getAccessFieldNum(const Instruction *I) const;
+
+  /// Obtain total number of instructions that access the struct fields
+  unsigned getTotalNumFieldAccess() const {
+    return LoadStoreFieldAccessMap.size() + CallInstFieldAccessMap.size();
+  }
+
+  /// Obtain execution count for the BasicBlock/Instruction from profiling info,
+  /// if any
+  /// %{
+  Optional<ExecutionCountType> getExecutionCount(const BasicBlock *BB) const {
+    return StructManager->getExecutionCount(BB);
+  }
+
+  Optional<ExecutionCountType> getExecutionCount(const Instruction *I) const {
+    return StructManager->getExecutionCount(I->getParent());
+  }
+  /// %}
+
+  /// Iterate through all Call/Invoke instructions that accesses a field and
+  /// summarize them into the function definitions
+  void summarizeFunctionCalls();
+
+  /// Calculate total hotness of all load/store field accesses
+  ExecutionCountType calculateTotalHotness() const;
+
+  /// Print all instructions that access any struct field
+  void debugPrintAllStructAccesses(raw_ostream &OS);
+
+  /// For stats
+  /// %{
+  void addStats(unsigned Category, unsigned Opcode = 0) {
+    StatCounts[Category]++;
+    if (Category == StructFieldAccessManager::DebugStats::DS_GepUnknownUse) {
+      if (UnknownOpcodes.find(Opcode) == UnknownOpcodes.end())
+        UnknownOpcodes[Opcode] = 0;
+      else
+        UnknownOpcodes[Opcode]++;
+    }
+  }
+
+  unsigned getStats(unsigned Category) const { return StatCounts[Category]; }
+  /// %}
+
+  void printUnknownOpcodes(raw_ostream &OS) const {
+    if (UnknownOpcodes.size() == 0)
+      return;
+    OS << "Unknown opcodes stats: \n";
+    for (auto &it : UnknownOpcodes) {
+      OS << "Opcode " << it.first << ": " << it.second << " times\n";
+    }
+  }
+
+private:
+  bool Eligiblity;
+  const Module &CurrentModule;
+  const StructType *StructureType;
+  const StructFieldAccessManager::StructDefinitionType StructDefinition;
+  const DICompositeType *DebugInfo;
+  unsigned NumElements;
+  const StructFieldAccessManager *StructManager;
+  /// For stats
+  std::vector<unsigned> StatCounts;
+  std::unordered_map<unsigned, unsigned> UnknownOpcodes;
+
+  /// A map records all load/store instructions accessing which field of the
+  /// structure
+  std::unordered_map<const Instruction *, unsigned> LoadStoreFieldAccessMap;
+
+  /// A map records all call/invoke instructions accessing which field of the
+  /// structure
+  std::unordered_map<const Instruction *, FunctionCallInfo *>
+      CallInstFieldAccessMap;
+
+  /// A map records all functions that has calls with field accesses and their
+  /// calling patterns
+  std::unordered_map<const Function *, FunctionAccessPattern *>
+      FunctionAccessMap;
+
+private:
```

Comment:
Easwaran and I decided to add this comment like lib/IR/ConstantsContext.h did to note the file is only private some cpp files.